enhancements/multi-arch/heterogenous-architecture.md

```diff
@@ -0,0 +1,270 @@
+
+### Upgrades - Cincinnati and CVO
+
+In addition to the single arch payload graph that we have today, [support][OCPPLAN-7642] needs to be added for an additional graph for hetero arch (called "multi" as noted above) which CVO will need to look at to
+make update recommendations. In order for CVO to know that the user wants to move to a heterogenous payload, a new field needs to be set in the cluster version resource or in the infrastructure resource for CVO to
+look at the hetero graph to make recommendations. There are a few upgrade scenarios to consider:
+
+- heterogenous -> heterogenous payload upgrade
+- homogenous -> heterogenous payload upgrade
+- heterogenous -> homogenous payload upgrade
+
+Because the upgrades can be from homogenous to heterogenous (and vice versa), in order to find the current release in the graph, the heterogenous graph would have both the homogenous and heterogenous release images.
+
+Cincinnati needs to [support][OCPPLAN-7643] parsing manifest list images to be able to upgrade to a heterogenous architecture payload.
+
+### Imagestreams
+
+Imagestreams need to better support manifest list images. As of today they do recognize manifest lists, but it only imports images matching the control plane architecture. The idea here would be to import the manifest listed image in its entirety.
+This also has other implications such as the fact that the internal registry does not support manifest listed images. So, with full manifest list support in imagestreams, pullthrough is not possible for manifest lists. We would ask users
+to use quay.io rather than use the internal registry if full manifest list support is needed.
+
+The oc command like debug, must-gather etc use the imagestreams and therefore would usee an x86 image to deploy on the worker nodes which would fail with an ARM worker.As a workaround, they can be provided the manifest listed image with
+the `--image` parameter which would allow the commands to work on the worker nodes as well.
+
+### OLM
+
+Operators need to specify which architectures they support based on annotations.There also needs to be a mechanism to ensure that the architecture is actually supported if that label is present. Also, all architectures should share the same
+content. Images potentially need to be available on architectures where there is no support for the operator as well? But the filtering should make sure only the supported arches show up.
+
+### Machine Config Operator
+
+No changes are needed in the MCO for Phase 1 as the machine-os-content image would be a manifest listed image and the machine config daemon would extract the relevant architecture's machine-os-content based on the node it runs on. In the future architecture
+specific configurations should not be required, instead they need to be templatized and be generalized for certain items, for example kubelet system reserved memory could scale based on page size. 
```

Comment:
so:
initially, we'll expect the user to provide any additional args/config in the machineset config
middle-state, we'll explicitly configure those things based on the arch
final-state, we'll auto-configure them based on generalized analysis

?

Reply:
i still have to understand from Colin/Mrunal the changes involved here for MCO.

Reply:
@mrunalp could you clarify what specifically would change in MCO when we support heterogeneous clusters a opposed to what we are doing today ?